app/src/main/java/com/github/HumanLearning2021/HumanLearningApp/firestore/FirestoreDatabaseService.kt

```diff
@@ -71,37 +95,146 @@ class FirestoreDatabaseService(
+        val documentRef = datasets.add(data).await()
+        return documentRef.get().await().toObject(DatasetSchema::class.java)!!.toPublic()
     }
 
-    override suspend fun getDataset(id: Any): Dataset? {
-        TODO("Not yet implemented")
+    override suspend fun getDataset(id: Any): FirestoreDataset? {
+        val ds = datasets.document(id as String).get().await().toObject(DatasetSchema::class.java)
+        return ds?.toPublic()
     }
 
     override suspend fun deleteDataset(id: Any) {
-        TODO("Not yet implemented")
+        datasets.document(id as String).get().addOnCompleteListener {
+            if (!it.isSuccessful) {
+                throw java.lang.IllegalArgumentException("Dataset with id $id is not contained in the databse")
+            }
+        }
+        datasets.document(id as String).delete().addOnFailureListener { Log.w(this.toString(), "Deleting dataset ${datasets.id} from ${this.db} failed") }
     }
 
     override suspend fun putRepresentativePicture(picture: Uri, category: Category) {
-        TODO("Not yet implemented")
+        require(category is FirestoreCategory)
+        val categoryRef = categories.document(category.id)
+        categoryRef.get()
+            .addOnFailureListener { throw java.lang.IllegalArgumentException("The database ${this.db} does not contain the category with ${category.id}") }
+        val imageRef = imagesDir.child("${UUID.randomUUID()}")
+        imageRef.putFile(picture).await()
+        val data =
+            PictureSchema(categories.document(category.id), "gs://${imageRef.bucket}/${imageRef.path}")
+        categoryRef.update("representativePicture", data.url).addOnFailureListener {
+            Log.w(
+                this.toString(),
+                "Setting representative picture of category ${category.id} to picture at ${data.url} failed"
+            )
+        }
     }
 
 
-    override fun getDatasets(): Set<Dataset> {
-        TODO("Not yet implemented")
+    @OptIn(ExperimentalStdlibApi::class)
+    override suspend fun getDatasets(): Set<FirestoreDataset> {
+        val ds = datasets.get().await().documents
+        return buildSet {
+            for (d in ds) {
+                val obj = d.toObject(DatasetSchema::class.java)
+                if (obj == null) {
+                    Log.w(this.toString(), "Failed to load dataset ${d.id}")
+                } else {
+                    add(obj.toPublic())
+                }
+            }
+        }
+    }
+
+    override suspend fun removeCategoryFromDataset(
+        dataset: Dataset,
+        category: Category
+    ): FirestoreDataset {
+        require(dataset is FirestoreDataset)
+        require(category is FirestoreCategory)
+
+        val dsCategories = dataset.categories
+        if (!dsCategories.contains(category))
+            throw IllegalArgumentException("The category ${category.id} is not contained in the dataset ${dataset.id}")
+
+        var res = dataset as FirestoreDataset
```

Comment:
Huge race condition there. just use `await`

Reply:
Is it though? Because here I access the field of an object not from the database.

Reply:
you return `res`. You cant guarantee that the Listener callback executes before the method returns, therefore race condition. This `res`-then-async-then-`return res` pattern is what `await()` is designed to do except without race conditions

Reply:
Oh okay, I understand now... I will fix this.